### DISCUSSION TITLE
Rust应该怎么导入

### DISCUSSION THREAD

**Discussion by pigeon-fancier** [MaaXYZ/MaaFramework#263] [Q&A] [answered]
复现步骤：
```shell
cargo new maa_xxx
```
Cargo.toml
```toml
[package]
name = "maa_xxx"
version = "0.1.0"
edition = "2021"

[dependencies]
maa-framework = "0.5.0"
```
```shell
cargo run



Caused by:
  process didn't exit successfully: `xxx\target\debug\build\maa-framework-da8f69a5e0001d68\build-script-build` (exit code: 101)
  --- stdout
  cargo:rerun-if-changed=./cmake/CMakeLists.txt

  --- stderr
  thread 'main' panicked at xxx\rust\cargo_home\registry\src\index.crates.io-6f17d22bba15001f\maa-framework-0.5.0\build.rs:19:9:
  Unable to find MaaFramework libraries
  stack backtrace:
     0: std::panicking::begin_panic_handler
               at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library\std\src\panicking.rs:652
     1: core::panicking::panic_fmt
               at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library\core\src\panicking.rs:72
     2: std::rt::lang_start
     3: std::rt::lang_start
     4: __ImageBase
     5: std::rt::lang_start
     6: std::rt::lang_start_internal
               at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library\std\src\rt.rs:141
     7: std::rt::lang_start
     8: main
     9: invoke_main
               at D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
    10: __scrt_common_main_seh
               at D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
    11: BaseThreadInitThunk
    12: RtlUserThreadStart
  note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
warning: build failed, waiting for other jobs to finish..
```

**Comment by KevinT3Hu** [comment, marked as answer]
你需要安装MaaFramework

推荐的方法是安装到你的本地。如果是Windows系统，解压最新版本的fw后将路径添加到CMake的search path中，具体方法查询CMake相关文档。最简单的方法是添加到PATH环境变量。

其次你可以添加maa_framework的download feature，会自动在构建时下载对应版本的fw，但这是不推荐的方法。

无论采取何种方法，最终的程序运行时都仍然需要fw的动态库。

**Comment by pigeon-fancier** [comment]
PATH环境变量已经添加win-x86_64.zip解压后的目录/bin
报错短了一点
error: failed to run custom build command for `maa-framework v0.5.0`

Caused by:
  process didn't exit successfully: `xxx\target\debug\build\maa-framework-da8f69a5e0001d68\build-script-build` (exit code: 101)
  --- stdout
  cargo:rerun-if-changed=./cmake/CMakeLists.txt

  --- stderr
  thread 'main' panicked at xxx\cargo_home\registry\src\index.crates.io-6f17d22bba15001f\maa-framework-0.5.0\build.rs:19:9:
  Unable to find MaaFramework libraries
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...

**Reply by pigeon-fancier** [reply]
@KevinT3Hu 辛苦大佬再看下

**Reply by KevinT3Hu** [reply]
不要添加/bin，应该添加安装的根目录

**Reply by pigeon-fancier** [reply]
> 不要添加/bin，应该添加安装的根目录
改成根目录后的报错

error: failed to run custom build command for `maa-framework v0.5.0`

Caused by:
  process didn't exit successfully: `xxx\target\debug\build\maa-framework-da8f69a5e0001d68\build-script-build` (exit code: 101)
  --- stdout
  cargo:rerun-if-changed=./cmake/CMakeLists.txt
  CMake Error at CMakeLists.txt:5 (find_package):
    Could not find a package configuration file named "MaaFramework.cmake"
    provided by package "MaaFramework".

    Add the installation prefix of "MaaFramework" to CMAKE_PREFIX_PATH or set
    "MaaFramework_DIR" to a directory containing one of the above files.  If
    "MaaFramework" provides a separate development package or SDK, be sure it
    has been installed.



  cargo:rustc-link-lib=MaaFramework
  cargo:rustc-link-lib=MaaToolkit
  cargo:rerun-if-env-changed=TARGET
  cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS_x86_64-pc-windows-msvc
  cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS_x86_64_pc_windows_msvc
  cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS
  cargo:rerun-if-changed=headers/maa_framework.h
  Unable to find libclang: "couldn't find any valid shared libraries matching: ['clang.dll', 'libclang.dll'], set the `LIBCLANG_PATH` environment variable to a path where one of these files can be found (invalid: [])"
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

**Reply by KevinT3Hu** [reply]
首先确认你的maafw安装目录下有share/cmake/MaaFramework文件夹和MaaFramework.cmake,MaaFrameworkConfig.cmake,MaaFramework-release.cmake这三个文件，然后clone maa-framework-rs这个repo，在repo内的cmake/文件夹下运行`cmake -B build . --debug-find`，把运行输出发出来。

**Reply by pigeon-fancier** [reply]
> 首先确认你的maafw安装目录下有share/cmake/MaaFramework文件夹和MaaFramework.cmake,MaaFrameworkConfig.cmake,MaaFramework-release.cmake这三个文件，然后clone maa-framework-rs这个repo，在repo内的cmake/文件夹下运行`cmake -B build . --debug-find`，把运行输出发出来。

cmake -B build . --debug-find
Running with debug output on for the `find` commands.
-- Selecting Windows SDK version 10.0.26100.0 to target Windows 10.0.22631.
CMake Debug Log at CMakeLists.txt:5 (find_package):
    D:/application/codeEnv/C/c_lib/maa-fw/share/cmake/MaaFramework/MaaFramework.cmake



IncludeDir=D:/application/codeEnv/C/c_lib/maa-fw/include
MaaFrameworkLibraries=D:/application/codeEnv/C/c_lib/maa-fw/lib/MaaFramework.lib
MaaToolkitLibraries=D:/application/codeEnv/C/c_lib/maa-fw/lib/MaaToolkit.lib
-- Configuring done (0.0s)
-- Generating done (0.1s)
-- Build files have been written to: D:/code/code-all/Rust/maa-framework-rs/cmake/build

**Reply by KevinT3Hu** [reply]
看起来你的fw配置没有问题。但是我确实忘记说bindgen需要libclang，参考[这里](https://rust-lang.github.io/rust-bindgen/requirements.html)配置。

配置好后可以分别在maa-framework-rs和你的crate里跑一下cargo build看看是否成功

**Reply by pigeon-fancier** [reply]
> 看起来你的fw配置没有问题。但是我确实忘记说bindgen需要libclang，参考[这里](https://rust-lang.github.io/rust-bindgen/requirements.html)配置。
> 
> 配置好后可以分别在maa-framework-rs和你的crate里跑一下cargo build看看是否成功

感谢大佬，解决了
